File: behavioral/typescript/visitor.ts
/*
 * VISITOR DESIGN PATTERN
 * 
 * WHAT IS IT?
 * The Visitor pattern lets you define a new operation without changing the classes of the elements on which it operates.
 * It separates the algorithm from the object structure it operates on.
 * 
 * WHY DOES IT EXIST?
 * - When you need to add operations to a complex object structure without changing the structure
 * - When you want to perform operations across different object types
 * - When you want to centralize related operations in a single class
 * 
 * WHEN TO USE IT?
 * - When you have a complex object structure with different types of elements
 * - When you need to perform many unrelated operations on these elements
 * - When the object structure is stable but you want to add new operations frequently
 * - Common use cases: document processing, compiler AST traversal, file system operations, shopping cart calculations
 * 
 * PROS:
 * - Open/Closed Principle (can add new operations without changing element classes)
 * - Single Responsibility Principle (moves related operations to a single visitor)
 * - Can accumulate state while traversing the object structure
 * - Works well with Composite pattern
 * 
 * CONS:
 * - Adding new element classes requires updating all visitors
 * - Can violate encapsulation by exposing element internals
 * - Can lead to complex visitor hierarchies
 * - May make the code harder to understand
 * 
 * REAL-WORLD ANALOGY:
 * Think of a tax inspector visiting different types of businesses. The inspector (visitor) visits a restaurant, a retail store, and a factory (elements). Each business type calculates taxes differently. The inspector knows how to calculate taxes for each type but doesn't change how the businesses operate. You can add new types of inspections (health inspector, fire inspector) without changing the businesses.
 */

// Visitor interface - declares visit methods for each concrete element
interface Visitor {
  visitBook(book: Book): void;
  visitFruit(fruit: Fruit): void;
  visitElectronic(electronic: Electronic): void;
}

// Element interface - declares the accept method
interface Item {
  accept(visitor: Visitor): void;
  getPrice(): number;
}

// Concrete Element 1: Book
class Book implements Item {
  private title: string;
  private author: string;
  private price: number;
  
  constructor(title: string, author: string, price: number) {
    this.title = title;
    this.author = author;
    this.price = price;
  }
  
  accept(visitor: Visitor): void {
    visitor.visitBook(this);
  }
  
  getPrice(): number {
    return this.price;
  }
  
  getTitle(): string { return this.title; }
  getAuthor(): string { return this.author; }
}

// Concrete Element 2: Fruit
class Fruit implements Item {
  private name: string;
  private price: number;
  private weight: number;
  
  constructor(name: string, price: number, weight: number) {
    this.name = name;
    this.price = price;
    this.weight = weight;
  }
  
  accept(visitor: Visitor): void {
    visitor.visitFruit(this);
  }
  
  getPrice(): number {
    return this.price;
  }
  
  getName(): string { return this.name; }
  getWeight(): number { return this.weight; }
}

// Concrete Element 3: Electronic
class Electronic implements Item {
  private model: string;
  private brand: string;
  private price: number;
  
  constructor(brand: string, model: string, price: number) {
    this.brand = brand;
    this.model = model;
    this.price = price;
  }
  
  accept(visitor: Visitor): void {
    visitor.visitElectronic(this);
  }
  
  getPrice(): number {
    return this.price;
  }
  
  getModel(): string { return this.model; }
  getBrand(): string { return this.brand; }
}

// Concrete Visitor 1: Price Calculator
class PriceCalculator implements Visitor {
  private totalPrice: number = 0;
  
  visitBook(book: Book): void {
    this.totalPrice += book.getPrice();
    console.log(`Book: ${book.getTitle()} - $${book.getPrice()}`);
  }
  
  visitFruit(fruit: Fruit): void {
    this.totalPrice += fruit.getPrice();
    console.log(`Fruit: ${fruit.getName()} - $${fruit.getPrice()}`);
  }
  
  visitElectronic(electronic: Electronic): void {
    this.totalPrice += electronic.getPrice();
    console.log(`Electronic: ${electronic.getBrand()} ${electronic.getModel()} - $${electronic.getPrice()}`);
  }
  
  getTotalPrice(): number {
    return this.totalPrice;
  }
}

// Concrete Visitor 2: Description Generator
class DescriptionGenerator implements Visitor {
  visitBook(book: Book): void {
    console.log(`Book: "${book.getTitle()}" by ${book.getAuthor()} - A great read!`);
  }
  
  visitFruit(fruit: Fruit): void {
    console.log(`Fruit: ${fruit.getName()} (${fruit.getWeight()}kg) - Fresh and healthy!`);
  }
  
  visitElectronic(electronic: Electronic): void {
    console.log(`Electronic: ${electronic.getBrand()} ${electronic.getModel()} - High-tech gadget!`);
  }
}

// Concrete Visitor 3: Tax Calculator
class TaxCalculator implements Visitor {
  private totalTax: number = 0;
  
  visitBook(book: Book): void {
    const tax = book.getPrice() * 0.05;  // 5% tax on books
    this.totalTax += tax;
    console.log(`Book tax: $${tax}`);
  }
  
  visitFruit(fruit: Fruit): void {
    const tax = fruit.getPrice() * 0.02;  // 2% tax on food
    this.totalTax += tax;
    console.log(`Fruit tax: $${tax}`);
  }
  
  visitElectronic(electronic: Electronic): void {
    const tax = electronic.getPrice() * 0.10;  // 10% tax on electronics
    this.totalTax += tax;
    console.log(`Electronic tax: $${tax}`);
  }
  
  getTotalTax(): number {
    return this.totalTax;
  }
}

// Object Structure - manages the elements
class ShoppingCart {
  private items: Item[] = [];
  
  addItem(item: Item): void {
    this.items.push(item);
  }
  
  acceptVisitor(visitor: Visitor): void {
    for (const item of this.items) {
      item.accept(visitor);
    }
  }
}

// Example 2: File System Visitor
// Shows how Visitor can be used for file system operations

interface FileSystemVisitor {
  visitFile(file: File): void;
  visitDirectory(directory: Directory): void;
}

interface FileSystemNode {
  accept(visitor: FileSystemVisitor): void;
  getName(): string;
  getSize(): number;
}

class File implements FileSystemNode {
  private name: string;
  private size: number;
  
  constructor(name: string, size: number) {
    this.name = name;
    this.size = size;
  }
  
  accept(visitor: FileSystemVisitor): void {
    visitor.visitFile(this);
  }
  
  getName(): string { return this.name; }
  getSize(): number { return this.size; }
}

class Directory implements FileSystemNode {
  private name: string;
  private children: FileSystemNode[] = [];
  
  constructor(name: string) {
    this.name = name;
  }
  
  addChild(child: FileSystemNode): void {
    this.children.push(child);
  }
  
  accept(visitor: FileSystemVisitor): void {
    visitor.visitDirectory(this);
  }
  
  getName(): string { return this.name; }
  getSize(): number {
    return this.children.reduce((total, child) => total + child.getSize(), 0);
  }
}

class SizeCalculator implements FileSystemVisitor {
  private totalSize: number = 0;
  
  visitFile(file: File): void {
    this.totalSize += file.getSize();
    console.log(`File: ${file.getName()} - ${file.getSize()} bytes`);
  }
  
  visitDirectory(directory: Directory): void {
    console.log(`Directory: ${directory.getName()} - ${directory.getSize()} bytes`);
    this.totalSize += directory.getSize();
  }
  
  getTotalSize(): number {
    return this.totalSize;
  }
}

class XMLExporter implements FileSystemVisitor {
  visitFile(file: File): void {
    console.log(`  <file name="${file.getName()}" size="${file.getSize()}"/>`);
  }
  
  visitDirectory(directory: Directory): void {
    console.log(`  <directory name="${directory.getName()}" size="${directory.getSize()}">`);
  }
}

// Example 3: Employee Salary Calculator
// Shows how Visitor can be used for salary calculations

interface SalaryVisitor {
  visitEmployee(employee: Employee): void;
  visitManager(manager: Manager): void;
  visitDeveloper(developer: Developer): void;
  visitDesigner(designer: Designer): void;
}

interface Employee {
  accept(visitor: SalaryVisitor): void;
  getName(): string;
  getBaseSalary(): number;
}

class Manager implements Employee {
  private name: string;
  private baseSalary: number;
  private bonus: number;
  
  constructor(name: string, salary: number, bonus: number) {
    this.name = name;
    this.baseSalary = salary;
    this.bonus = bonus;
  }
  
  accept(visitor: SalaryVisitor): void {
    visitor.visitManager(this);
  }
  
  getName(): string { return this.name; }
  getBaseSalary(): number { return this.baseSalary; }
  getBonus(): number { return this.bonus; }
}

class Developer implements Employee {
  private name: string;
  private baseSalary: number;
  private overtimeHours: number;
  
  constructor(name: string, salary: number, overtime: number) {
    this.name = name;
    this.baseSalary = salary;
    this.overtimeHours = overtime;
  }
  
  accept(visitor: SalaryVisitor): void {
    visitor.visitDeveloper(this);
  }
  
  getName(): string { return this.name; }
  getBaseSalary(): number { return this.baseSalary; }
  getOvertimeHours(): number { return this.overtimeHours; }
}

class Designer implements Employee {
  private name: string;
  private baseSalary: number;
  private projectsCompleted: number;
  
  constructor(name: string, salary: number, projects: number) {
    this.name = name;
    this.baseSalary = salary;
    this.projectsCompleted = projects;
  }
  
  accept(visitor: SalaryVisitor): void {
    visitor.visitDesigner(this);
  }
  
  getName(): string { return this.name; }
  getBaseSalary(): number { return this.baseSalary; }
  getProjectsCompleted(): number { return this.projectsCompleted; }
}

class BaseEmployee implements Employee {
  private name: string;
  private baseSalary: number;
  
  constructor(name: string, salary: number) {
    this.name = name;
    this.baseSalary = salary;
  }
  
  accept(visitor: SalaryVisitor): void {
    visitor.visitEmployee(this);
  }
  
  getName(): string { return this.name; }
  getBaseSalary(): number { return this.baseSalary; }
}

class SalaryCalculator implements SalaryVisitor {
  private totalSalary: number = 0;
  
  visitEmployee(employee: Employee): void {
    const salary = employee.getBaseSalary();
    this.totalSalary += salary;
    console.log(`Employee: ${employee.getName()} - $${salary}`);
  }
  
  visitManager(manager: Manager): void {
    const salary = manager.getBaseSalary() + manager.getBonus();
    this.totalSalary += salary;
    console.log(`Manager: ${manager.getName()} - $${salary} (base: $${manager.getBaseSalary()} + bonus: $${manager.getBonus()})`);
  }
  
  visitDeveloper(developer: Developer): void {
    const overtimePay = developer.getOvertimeHours() * 50.0;
    const salary = developer.getBaseSalary() + overtimePay;
    this.totalSalary += salary;
    console.log(`Developer: ${developer.getName()} - $${salary} (base: $${developer.getBaseSalary()} + overtime: $${overtimePay})`);
  }
  
  visitDesigner(designer: Designer): void {
    const projectBonus = designer.getProjectsCompleted() * 500.0;
    const salary = designer.getBaseSalary() + projectBonus;
    this.totalSalary += salary;
    console.log(`Designer: ${designer.getName()} - $${salary} (base: $${designer.getBaseSalary()} + project bonus: $${projectBonus})`);
  }
  
  getTotalSalary(): number {
    return this.totalSalary;
  }
}

// Demo code
console.log('=== Visitor Pattern Demo ===\n');

// Example 1: Shopping Cart
console.log('--- Shopping Cart Example ---');

const cart = new ShoppingCart();
cart.addItem(new Book('Design Patterns', 'Erich Gamma', 49.99));
cart.addItem(new Fruit('Apple', 1.99, 0.5));
cart.addItem(new Electronic('Apple', 'iPhone 15', 999.99));

console.log('\nCalculating total price:');
const priceCalc = new PriceCalculator();
cart.acceptVisitor(priceCalc);
console.log(`Total: $${priceCalc.getTotalPrice()}`);

console.log('\nGenerating descriptions:');
const descGen = new DescriptionGenerator();
cart.acceptVisitor(descGen);

console.log('\nCalculating tax:');
const taxCalc = new TaxCalculator();
cart.acceptVisitor(taxCalc);
console.log(`Total tax: $${taxCalc.getTotalTax()}`);

// Example 2: File System
console.log('\n--- File System Example ---');

const root = new Directory('root');
const file1 = new File('document.txt', 1024);
const file2 = new File('image.jpg', 2048);
const subdir = new Directory('subdir');
const file3 = new File('data.csv', 512);

subdir.addChild(file3);
root.addChild(file1);
root.addChild(file2);
root.addChild(subdir);

console.log('\nCalculating sizes:');
const sizeCalc = new SizeCalculator();
root.accept(sizeCalc);
console.log(`Total size: ${sizeCalc.getTotalSize()} bytes`);

console.log('\nExporting to XML:');
console.log('<filesystem>');
const xmlExporter = new XMLExporter();
root.accept(xmlExporter);
console.log('</filesystem>');

// Example 3: Employee Salary
console.log('\n--- Employee Salary Example ---');

const employees: Employee[] = [
  new Manager('Alice', 80000, 20000),
  new Developer('Bob', 70000, 10),
  new Designer('Charlie', 65000, 5)
];

console.log('\nCalculating total salary:');
const salaryCalc = new SalaryCalculator();
for (const emp of employees) {
  emp.accept(salaryCalc);
}
console.log(`Total salary: $${salaryCalc.getTotalSalary()}`);

File: behavioral/dart/visitor.dart
/*
 * VISITOR DESIGN PATTERN
 * 
 * WHAT IS IT?
 * The Visitor pattern lets you define a new operation without changing the classes of the elements on which it operates.
 * It separates the algorithm from the object structure it operates on.
 * 
 * WHY DOES IT EXIST?
 * - When you need to add operations to a complex object structure without changing the structure
 * - When you want to perform operations across different object types
 * - When you want to centralize related operations in a single class
 * 
 * WHEN TO USE IT?
 * - When you have a complex object structure with different types of elements
 * - When you need to perform many unrelated operations on these elements
 * - When the object structure is stable but you want to add new operations frequently
 * - Common use cases: document processing, compiler AST traversal, file system operations, shopping cart calculations
 * 
 * PROS:
 * - Open/Closed Principle (can add new operations without changing element classes)
 * - Single Responsibility Principle (moves related operations to a single visitor)
 * - Can accumulate state while traversing the object structure
 * - Works well with Composite pattern
 * 
 * CONS:
 * - Adding new element classes requires updating all visitors
 * - Can violate encapsulation by exposing element internals
 * - Can lead to complex visitor hierarchies
 * - May make the code harder to understand
 * 
 * REAL-WORLD ANALOGY:
 * Think of a tax inspector visiting different types of businesses. The inspector (visitor) visits a restaurant, a retail store, and a factory (elements). Each business type calculates taxes differently. The inspector knows how to calculate taxes for each type but doesn't change how the businesses operate. You can add new types of inspections (health inspector, fire inspector) without changing the businesses.
 */

// Visitor interface - declares visit methods for each concrete element
abstract class Visitor {
  void visitBook(Book book);
  void visitFruit(Fruit fruit);
  void visitElectronic(Electronic electronic);
}

// Element interface - declares the accept method
abstract class Item {
  void accept(Visitor visitor);
  double getPrice();
}

// Concrete Element 1: Book
class Book extends Item {
  final String _title;
  final String _author;
  final double _price;
  
  Book(this._title, this._author, this._price);
  
  @override
  void accept(Visitor visitor) {
    visitor.visitBook(this);
  }
  
  @override
  double getPrice() => _price;
  
  String get title => _title;
  String get author => _author;
}

// Concrete Element 2: Fruit
class Fruit extends Item {
  final String _name;
  final double _price;
  final double _weight;
  
  Fruit(this._name, this._price, this._weight);
  
  @override
  void accept(Visitor visitor) {
    visitor.visitFruit(this);
  }
  
  @override
  double getPrice() => _price;
  
  String get name => _name;
  double get weight => _weight;
}

// Concrete Element 3: Electronic
class Electronic extends Item {
  final String _model;
  final String _brand;
  final double _price;
  
  Electronic(this._brand, this._model, this._price);
  
  @override
  void accept(Visitor visitor) {
    visitor.visitElectronic(this);
  }
  
  @override
  double getPrice() => _price;
  
  String get model => _model;
  String get brand => _brand;
}

// Concrete Visitor 1: Price Calculator
class PriceCalculator implements Visitor {
  double _totalPrice = 0;
  
  @override
  void visitBook(Book book) {
    _totalPrice += book.getPrice();
    print('Book: ${book.title} - \$${book.getPrice()}');
  }
  
  @override
  void visitFruit(Fruit fruit) {
    _totalPrice += fruit.getPrice();
    print('Fruit: ${fruit.name} - \$${fruit.getPrice()}');
  }
  
  @override
  void visitElectronic(Electronic electronic) {
    _totalPrice += electronic.getPrice();
    print('Electronic: ${electronic.brand} ${electronic.model} - \$${electronic.getPrice()}');
  }
  
  double get totalPrice => _totalPrice;
}

// Concrete Visitor 2: Description Generator
class DescriptionGenerator implements Visitor {
  @override
  void visitBook(Book book) {
    print('Book: "${book.title}" by ${book.author} - A great read!');
  }
  
  @override
  void visitFruit(Fruit fruit) {
    print('Fruit: ${fruit.name} (${fruit.weight}kg) - Fresh and healthy!');
  }
  
  @override
  void visitElectronic(Electronic electronic) {
    print('Electronic: ${electronic.brand} ${electronic.model} - High-tech gadget!');
  }
}

// Concrete Visitor 3: Tax Calculator
class TaxCalculator implements Visitor {
  double _totalTax = 0;
  
  @override
  void visitBook(Book book) {
    final tax = book.getPrice() * 0.05;  // 5% tax on books
    _totalTax += tax;
    print('Book tax: \$${tax}');
  }
  
  @override
  void visitFruit(Fruit fruit) {
    final tax = fruit.getPrice() * 0.02;  // 2% tax on food
    _totalTax += tax;
    print('Fruit tax: \$${tax}');
  }
  
  @override
  void visitElectronic(Electronic electronic) {
    final tax = electronic.getPrice() * 0.10;  // 10% tax on electronics
    _totalTax += tax;
    print('Electronic tax: \$${tax}');
  }
  
  double get totalTax => _totalTax;
}

// Object Structure - manages the elements
class ShoppingCart {
  final List<Item> _items = [];
  
  void addItem(Item item) {
    _items.add(item);
  }
  
  void acceptVisitor(Visitor visitor) {
    for (final item in _items) {
      item.accept(visitor);
    }
  }
}

// Example 2: File System Visitor
// Shows how Visitor can be used for file system operations

abstract class FileSystemVisitor {
  void visitFile(File file);
  void visitDirectory(Directory directory);
}

abstract class FileSystemNode {
  void accept(FileSystemVisitor visitor);
  String getName();
  int getSize();
}

class File implements FileSystemNode {
  final String _name;
  final int _size;
  
  File(this._name, this._size);
  
  @override
  void accept(FileSystemVisitor visitor) {
    visitor.visitFile(this);
  }
  
  @override
  String getName() => _name;
  
  @override
  int getSize() => _size;
}

class Directory implements FileSystemNode {
  final String _name;
  final List<FileSystemNode> _children = [];
  
  Directory(this._name);
  
  void addChild(FileSystemNode child) {
    _children.add(child);
  }
  
  @override
  void accept(FileSystemVisitor visitor) {
    visitor.visitDirectory(this);
  }
  
  @override
  String getName() => _name;
  
  @override
  int getSize() {
    return _children.fold(0, (total, child) => total + child.getSize());
  }
}

class SizeCalculator implements FileSystemVisitor {
  int _totalSize = 0;
  
  @override
  void visitFile(File file) {
    _totalSize += file.getSize();
    print('File: ${file.getName()} - ${file.getSize()} bytes');
  }
  
  @override
  void visitDirectory(Directory directory) {
    print('Directory: ${directory.getName()} - ${directory.getSize()} bytes');
    _totalSize += directory.getSize();
  }
  
  int get totalSize => _totalSize;
}

class XMLExporter implements FileSystemVisitor {
  @override
  void visitFile(File file) {
    print('  <file name="${file.getName()}" size="${file.getSize()}"/>');
  }
  
  @override
  void visitDirectory(Directory directory) {
    print('  <directory name="${directory.getName()}" size="${directory.getSize()}">');
  }
}

// Example 3: Employee Salary Calculator
// Shows how Visitor can be used for salary calculations

abstract class SalaryVisitor {
  void visitEmployee(Employee employee);
  void visitManager(Manager manager);
  void visitDeveloper(Developer developer);
  void visitDesigner(Designer designer);
}

abstract class Employee {
  void accept(SalaryVisitor visitor);
  String getName();
  double getBaseSalary();
}

class Manager extends Employee {
  final String _name;
  final double _baseSalary;
  final double _bonus;
  
  Manager(this._name, this._baseSalary, this._bonus);
  
  @override
  void accept(SalaryVisitor visitor) {
    visitor.visitManager(this);
  }
  
  @override
  String getName() => _name;
  
  @override
  double getBaseSalary() => _baseSalary;
  
  double get bonus => _bonus;
}

class Developer extends Employee {
  final String _name;
  final double _baseSalary;
  final int _overtimeHours;
  
  Developer(this._name, this._baseSalary, this._overtimeHours);
  
  @override
  void accept(SalaryVisitor visitor) {
    visitor.visitDeveloper(this);
  }
  
  @override
  String getName() => _name;
  
  @override
  double getBaseSalary() => _baseSalary;
  
  int get overtimeHours => _overtimeHours;
}

class Designer extends Employee {
  final String _name;
  final double _baseSalary;
  final int _projectsCompleted;
  
  Designer(this._name, this._baseSalary, this._projectsCompleted);
  
  @override
  void accept(SalaryVisitor visitor) {
    visitor.visitDesigner(this);
  }
  
  @override
  String getName() => _name;
  
  @override
  double getBaseSalary() => _baseSalary;
  
  int get projectsCompleted => _projectsCompleted;
}

class BaseEmployee extends Employee {
  final String _name;
  final double _baseSalary;
  
  BaseEmployee(this._name, this._baseSalary);
  
  @override
  void accept(SalaryVisitor visitor) {
    visitor.visitEmployee(this);
  }
  
  @override
  String getName() => _name;
  
  @override
  double getBaseSalary() => _baseSalary;
}

class SalaryCalculator implements SalaryVisitor {
  double _totalSalary = 0;
  
  @override
  void visitEmployee(Employee employee) {
    final salary = employee.getBaseSalary();
    _totalSalary += salary;
    print('Employee: ${employee.getName()} - \$$salary');
  }
  
  @override
  void visitManager(Manager manager) {
    final salary = manager.getBaseSalary() + manager.bonus;
    _totalSalary += salary;
    print('Manager: ${manager.getName()} - \$$salary (base: \$${manager.getBaseSalary()} + bonus: \$${manager.bonus})');
  }
  
  @override
  void visitDeveloper(Developer developer) {
    final overtimePay = developer.overtimeHours * 50.0;
    final salary = developer.getBaseSalary() + overtimePay;
    _totalSalary += salary;
    print('Developer: ${developer.getName()} - \$$salary (base: \$${developer.getBaseSalary()} + overtime: \$$overtimePay)');
  }
  
  @override
  void visitDesigner(Designer designer) {
    final projectBonus = designer.projectsCompleted * 500.0;
    final salary = designer.getBaseSalary() + projectBonus;
    _totalSalary += salary;
    print('Designer: ${designer.getName()} - \$$salary (base: \$${designer.getBaseSalary()} + project bonus: \$$projectBonus)');
  }
  
  double get totalSalary => _totalSalary;
}

// Demo code
void main() {
  print('=== Visitor Pattern Demo ===\n');
  
  // Example 1: Shopping Cart
  print('--- Shopping Cart Example ---');
  
  final cart = ShoppingCart();
  cart.addItem(Book('Design Patterns', 'Erich Gamma', 49.99));
  cart.addItem(Fruit('Apple', 1.99, 0.5));
  cart.addItem(Electronic('Apple', 'iPhone 15', 999.99));
  
  print('\nCalculating total price:');
  final priceCalc = PriceCalculator();
  cart.acceptVisitor(priceCalc);
  print('Total: \$${priceCalc.totalPrice}');
  
  print('\nGenerating descriptions:');
  final descGen = DescriptionGenerator();
  cart.acceptVisitor(descGen);
  
  print('\nCalculating tax:');
  final taxCalc = TaxCalculator();
  cart.acceptVisitor(taxCalc);
  print('Total tax: \$${taxCalc.totalTax}');
  
  // Example 2: File System
  print('\n--- File System Example ---');
  
  final root = Directory('root');
  final file1 = File('document.txt', 1024);
  final file2 = File('image.jpg', 2048);
  final subdir = Directory('subdir');
  final file3 = File('data.csv', 512);
  
  subdir.addChild(file3);
  root.addChild(file1);
  root.addChild(file2);
  root.addChild(subdir);
  
  print('\nCalculating sizes:');
  final sizeCalc = SizeCalculator();
  root.accept(sizeCalc);
  print('Total size: ${sizeCalc.totalSize} bytes');
  
  print('\nExporting to XML:');
  print('<filesystem>');
  final xmlExporter = XMLExporter();
  root.accept(xmlExporter);
  print('</filesystem>');
  
  // Example 3: Employee Salary
  print('\n--- Employee Salary Example ---');
  
  final employees = <Employee>[
    Manager('Alice', 80000, 20000),
    Developer('Bob', 70000, 10),
    Designer('Charlie', 65000, 5)
  ];
  
  print('\nCalculating total salary:');
  final salaryCalc = SalaryCalculator();
  for (final emp in employees) {
    emp.accept(salaryCalc);
  }
  print('Total salary: \$${salaryCalc.totalSalary}');
}
